### ISSUE TITLE
checksum error jdk

### ISSUE THREAD

**Issue by vsoch** [LLNL/spack#27666]
### Steps to reproduce

```console
spack install bowtie
```

### Error message

```console
==> Error: ChecksumError: sha256 checksum failed for /tmp/runner/spack-stage/spack-stage-jdk-14_36-j2k4e757xfunfadrcvs7z7hgsrcxygbp/jdk-14_linux-x64_bin.tar.gz
    Expected 4639bbaecc9cc606f1a4b99fda1efcaefcbf57a7025b3828b095093a6c866afd but got 80212f7748c2671b89b6085000717c747851db004409dce3bb97f3a2aeb91cdd

/opt/spack/lib/spack/spack/package.py:1390, in do_fetch:
       1387        self.stage.fetch(mirror_only, err_msg=err_msg)
       1388        self._fetch_time = time.time() - start_time
       1389
  >>   1390        if checksum and self.version in self.versions:
       1391            self.stage.check()
       1392
       1393        self.stage.cache_local()
```

### Information on your system

Github workflow failure

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
I could reproduce. Apparently the method we used to accept Oracle's license doesn't work anymore. The downloaded artifact contains some `html` for an "Unauthorized request".

**Comment by Pandapip1** [comment]
Can repro. Copying over the kludge I sent to my slack:

If you've had issues installing some spack packages, I may have found a workaround. Here's an example error log:

```shell
==> Error: FetchError: All fetchers failed for spack-stage-jdk-14_36-etpz5i6nxtu6esqnmd2wjjefof26eiji

/ref/djslab/software/spack-0.18.1/lib/spack/spack/package.py:1537, in do_fetch:
       1534
       1535        self.stage.create()
       1536        err_msg = None if not self.manual_download else self.download_instr
  >>   1537        start_time = time.time()
       1538        self.stage.fetch(mirror_only, err_msg=err_msg)
       1539        self._fetch_time = time.time() - start_time
       1540
```

The reason for this error is that Oracle, the company that owns Java, now requires an authentication token to download Java, meaning that anything that depends on the jdk package can no longer be installed. A temporary workaround is to update the sources to allow installing the openjdk instead:

1. Navigate to `spack/var/spack/repos/builtin/packages/jdk`
2. Open `package.py` in your preferred editor.
3. Below the lines that say:
  ```python
# To add the latest version, go to the homepage listed above,
# click "JDK Download", click "Accept License Agreement", right-click the
# Linux .tar.gz link, and select Copy Link Address. The checksum can be
# found in a link above. The build number can be deciphered from the URL.
# Alternatively, run `bin/java -version` after extracting. Replace '+'
# symbol in version with '_', otherwise it will be interpreted as a variant
  ```
  Add the following line. Make sure to indent it properly.
  ```python
version('14.0.1_openjdk', sha256='22ce248e0bd69f23028625bede9d1b3080935b68d011eaaf9e241f84d6b9c4cc', url='https://download.java.net/java/GA/jdk14.0.1/664493ef4a6946b186ff29eb326336a2/7/GPL/openjdk-14.0.1_linux-x64_bin.tar.gz')
  ```
4. Find the following function:
  ```python
@classmethod
def determine_version(cls, exe):
    output = Executable(exe)('-version', output=str, error=str)

    # Make sure this is actually Oracle JDK, not OpenJDK
    if 'openjdk' in output:
        return None

    match = re.search(r'\(build (\S+)\)', output)
    return match.group(1).replace('+', '_') if match else None
  ```
  Comment out the openjdk detection:
  ```python
# Make sure this is actually Oracle JDK, not OpenJDK
#if 'openjdk' in output:
#    return None
  ```
6. Install `jdk@14.0.1_openjdk`

You might be able to install the packages that were previously failing. You only need to do this once per lab.
Note that OpenJDK and Oracle JDK are not 100% compatible with one another, so you may run into issues.

**Comment by bernhardkaindl** [comment]
fixed by #43264
